prov/rxm/src/rxm_ep.c

```diff
@@ -967,17 +967,71 @@ rxm_ep_sar_tx_prepare_segment(struct rxm_ep *rxm_ep, struct rxm_conn *rxm_conn,
 	return tx_buf;
 }
 
+static inline ssize_t
+rxm_ep_sar_tx_prepare_and_send_segment(struct rxm_ep *rxm_ep, struct rxm_conn *rxm_conn,
+				       size_t data_len, size_t *remain_len, size_t seg_len,
+				       size_t seg_no, size_t offset, uint64_t data,
+				       uint64_t flags, uint64_t tag, uint64_t comp_flags,
+				       uint8_t op, enum rxm_sar_seg_type seg_type,
+				       struct rxm_tx_entry *tx_entry)
+{
+	struct rxm_tx_buf *tx_buf;
+	int send_failed = 0;
+	ssize_t ret;
+
+	tx_buf = rxm_ep_sar_tx_prepare_segment(rxm_ep, rxm_conn, data_len,
+					       seg_len, seg_no, offset, data,
+					       flags, tag, comp_flags, op,
+					       seg_type, tx_entry);
+	if (OFI_UNLIKELY(!tx_buf)) {
+		tx_entry->msg_id = UINT64_MAX;
+		if (seg_type == RXM_SAR_SEG_FIRST) {
+			/* if TX buffer allocation for the first segment fails,
+			 * release TX entry and report to user */
+			rxm_tx_entry_release(rxm_conn->send_queue, tx_entry);
+		}
+		while (!dlist_empty(&tx_entry->deferred_tx_buf_list)) {
+			dlist_pop_front(&tx_entry->deferred_tx_buf_list,
+					struct rxm_tx_buf, tx_buf, hdr.entry);
+			rxm_tx_buf_release(tx_entry->ep, tx_buf);
+		}
+		return -FI_EAGAIN;
+	}
+
+	if (!send_failed) {
+		ret = fi_send(rxm_conn->msg_ep, &tx_buf->pkt,
+			      sizeof(struct rxm_pkt) +
+			      tx_buf->pkt.ctrl_hdr.seg_size,
+			      tx_buf->hdr.desc, 0, tx_buf);
+		if (OFI_UNLIKELY(ret)) {
+			if (seg_type == RXM_SAR_SEG_FIRST) {
+				/* if the sending for the first segment fails,
+				 * release resources and report this to user */
+				rxm_tx_buf_release(tx_entry->ep, tx_buf);
+				rxm_tx_entry_release(rxm_conn->send_queue, tx_entry);
+				return -FI_EAGAIN;
+			}
+			send_failed = 1;
```

Comment:
Setting and checking this variable would have no effect since it's been moved out of the while loop.

Reply:
Done. Thanks!